l3kernel/l3expan.dtx

```diff
@@ -125,87 +130,92 @@
-%   \cs[no-index]{exp_args:N\meta{variant}} function needed to carry out the
-%   expansion. There is no need to re-apply \cs{cs_generate_variant:Nn} after
-%   changing the definition of the parent function: the variant will always
-%   use the current definition of the parent. Providing variants repeatedly is
-%   safe as \cs{cs_generate_variant:Nn} will only create new definitions if
-%   there is not already one available.
-%
-%   Only |n|~and |N| arguments can be changed to other types.  The only
-%   allowed changes are
+%   This function is used to define argument specifier variants of the
+%   base function \cs[no-index]{\meta{name}:\meta{base arg spec}}.
+%   The head of \meta{base arg spec} must consist of |N|, |n| or |p|.
+%   The \meta{variant argument specifiers} is a comma-separated list of
+%   argument specifiers used to build \meta{variant arg spec}'s,
+%   by mapping the heading letters of \meta{base arg spec}.
+%   The only possible mappings are
 %   \begin{itemize}
-%   \item |c|~variant of an |N|~parent;
-%   \item |o|, |V|, |v|, |f|, |e|, or~|x| variant of an |n|~parent;
-%   \item |N|, |n|, |T|, |F|, or |p| argument unchanged.
+%   \item |N| to |c|;
+%   \item |n| to |o|, |V|, |v|, |f|, |e|, or |x|;
+%   \item |p| to |p|;
 %   \end{itemize}
-%   This means the \meta{parent} of a \meta{variant} form is always
-%   unambiguous, even in cases where both an |n|-type parent and an
-%   |N|-type parent exist, such as for \cs{tl_count:n} and
-%   \cs{tl_count:N}.
-%
-%   When creating variants for conditional functions,
-%   \cs{prg_generate_conditional_variant:Nnn} provides a convenient way
-%   of handling the related function set.
+%   with the restriction that |p| arguments must be followed
+%   \emph{in fine} by a braced group.
+%
+%   If it does not exist, each variant function
+%   \cs[no-index]{\meta{name}|:|\meta{variant arg spec}} is globally
+%   created, as is the function \cs[no-index]{exp_args:N\meta{variant arg spec}}
+%   needed to expand the arguments as necessary before they
+%   are passed to the base function.
+%   So for example
+%   \begin{quote}\ttfamily
+%     |\cs_set:Npn \foo:Nn #1#2 {|\meta{code here}|}|\\
+%     |\cs_generate_variant:Nn \foo:Nn { c }|
+%   \end{quote}
+%   creates a new function |\foo:cn| which expands its first
+%   argument into a control sequence further passed as first argument to
+%   |\foo:Nn|. Similarly
+%   \begin{quote}
+%     |\cs_generate_variant:Nn \foo:Nn { NV , cV }|
+%   \end{quote}
+%   generates the variants |\foo:NV| and |\foo:cV| in the same
+%   way. Notice that
+%   \begin{quote}
+%     |\cs_generate_variant:Nn \foo:cn { cV }|
+%   \end{quote}
+%   is an error to generate |\foo:cV| because |cn| is not a
+%   \meta{base arg spec}. On the contrary
+%   \begin{quote}
+%     |\cs_generate_variant:Nn \foo:nc { Vc }|
+%   \end{quote}
+%   is allowed because |c| is not part of the \meta{base arg spec} head.
```

Comment:
The first works because it is like `\cs_generate_variant:Nn \foo:nc  { V }`
More fun with
```
\cs_new:Npn \foo:nwhatever {}
\cs_generate_variant:Nn \foo:nwhatever { Vwhat }
\show \foo:Vwhatever
```
The first argument of `\cs_generate_variant:Nn` is a "base function", and this notion has never been defined properly. One definition that appeared in discussions is that the signature of base functions must be `:[nN]+T?F?`. Another possibility is that the signature must be `:[npN]+.*`, which seems to correspond to the actual implementation (to be verified). Next example works,
```
\cs_new:Npn \foo:npnpn {}
\cs_generate_variant:Nn \foo:npnpn { Vpvpe }
\show \foo: Vpvpe
\show \exp_args:NVpvpe
```